plugins/experimental/carp/carp.cc

```diff
@@ -0,0 +1,713 @@
+      string status;
+      g_CarpConfigPool->getGlobalHashAlgo()->dump(status);
+      TSHttpTxnSetHttpRetStatus(txnp, TS_HTTP_STATUS_MULTI_STATUS);
+      TSHttpTxnErrorBodySet(txnp, TSstrdup(status.c_str()), status.length(), NULL);
+      return -1;
+    }
+    TSDebug(DEBUG_TAG_HOOK, "Found %s header with unknown value of %s, ignoring", CARP_ROUTED_HEADER.c_str(), value.c_str());
+    removeHeader(bufp, hdr_loc, CARP_ROUTED_HEADER);
+  }
+  return 1; // all OK
+}
+
+static bool
+checkListForSelf(std::vector<HashNode *> list)
+{
+  for (size_t k = 0; k < list.size(); k++) {
+    if (list[k]->isSelf) return true;
+  }
+  return false;
+}
+
+/**
+ bIsPOSTRemap = false --- Hash request and forward to peer
+ bIsPOSTRemap = true --- hash request, extract OS sockaddr, insert forwarding header, forward
+ */
+static int
+handleRequestProcessing(TSCont contp, TSEvent event, void *edata, bool bIsPOSTRemap)
+{
+  TSHttpTxn txnp = (TSHttpTxn) edata;
+  TSMBuffer bufp;
+  TSMLoc hdr_loc;
+  TSMLoc url_loc;
+
+  // get the client request so we can get URL and add header
+  if (TSHttpTxnClientReqGet(txnp, &bufp, &hdr_loc) != TS_SUCCESS) {
+    TSError("carp couldn't get request headers");
+    return -1;
+  }
+
+  int method_len;
+  const char *method = TSHttpHdrMethodGet(bufp, hdr_loc, &method_len);
+  if (NULL == method) {
+    TSError("carp couldn't get http method");
+    TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+    return -1;
+  }
+  if (((method_len == TS_HTTP_LEN_DELETE) && (strncasecmp(method, TS_HTTP_METHOD_DELETE, TS_HTTP_LEN_DELETE) == 0)) ||
+      ((method_len == TS_HTTP_LEN_PURGE) && (strncasecmp(method, TS_HTTP_METHOD_PURGE, TS_HTTP_LEN_PURGE) == 0))) {
+    TSDebug(DEBUG_TAG_HOOK, "Request method is '%s' so not routing request", string(method,method_len).c_str());
+    TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+    return 0;
+  }
+
+  if (TSHttpHdrUrlGet(bufp, hdr_loc, &url_loc) != TS_SUCCESS) {
+    TSError("carp couldn't get url");
+    TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+    return -1;
+  }
+  // if has carp loop prevention header, do not remap
+  int iTemp = processCarpRoutedHeader(txnp, bufp, hdr_loc);
+  if(iTemp <= 0) { // if dump or do not remap
+    // check origin client request's scheme for premap mode
+    if (!bIsPOSTRemap) {
+      string oriScheme;
+      if (!getHeader(bufp, hdr_loc, CARP_PREMAP_SCHEME, oriScheme)) {
+        TSDebug(DEBUG_TAG_HOOK, "couldn't get '%s' header", CARP_PREMAP_SCHEME.c_str());
+      } else {
+        bool isHttps = (oriScheme == TS_URL_SCHEME_HTTPS);
+        
+        if (isHttps) {
+          TSUrlSchemeSet(bufp, url_loc, TS_URL_SCHEME_HTTPS, TS_URL_LEN_HTTPS);
+        } else {
+          TSUrlSchemeSet(bufp, url_loc, TS_URL_SCHEME_HTTP, TS_URL_LEN_HTTP);  
+        }   
+        
+	removeHeader(bufp, hdr_loc, CARP_STATUS_HEADER);
+        removeHeader(bufp, hdr_loc, CARP_ROUTED_HEADER);
```

Comment:
Will do. I'm going to do a big clang-format of this after addressing your comments because the one I thought I did doesn't seem to have worked.